src/components/DonutChart.js

```diff
@@ -61,101 +67,209 @@ const DonutChart = React.createClass({
 
   getInitialState () {
     return {
-      activeIndex: -1
+      activeIndex: this.props.activeIndex || -1
     };
   },
 
+  componentWillMount () {
+    this._setupD3Functions(this.props);
+  },
+
   componentDidMount () {
-    this.setState({
-      activeIndex: this.props.activeIndex
-    });
+    this._animateChart();
   },
 
   componentWillReceiveProps (newProps) {
+    if (!_isEqual(this.props.data, newProps.data)) {
+      this._setupD3Functions(newProps);
+      this._animateChart();
+    }
+
     if (newProps.activeIndex !== this.props.activeIndex) {
       this.setState({
         activeIndex: newProps.activeIndex
       });
+
+      this._animateActiveArc(this.props.activeIndex, newProps.activeIndex);
+    }
+  },
+
+  shouldComponentUpdate (nextProps, nextState) {
+    return !_isEqual(this.props, nextProps) || !_isEqual(this.state, nextState);
+  },
+
+  _setupD3Functions (props) {
+    const dataSets = props.data.map(item => {
+      return item.value;
+    });
+    const valueTotal = dataSets.length ? dataSets.reduce((a, b) => {
+      return a + b;
+    }) : 0;
```

Comment:
That is pretty cool and very good info.  I didn't know you could do that.  In this case though I need the results of the map and the results of the reduce to be separate.  I only use the reduce results to determine the end angle of the chart where the map results are used to determine the values of each arc in the chart.